### PR TITLE
Implement `vela logs`

### PR DESCRIPTION
**Pull request by zzxwill**
Based on modified `stern` project, tail logs of pods
of an application

## About this draft release
This PR is intended to implement https://github.com/cloud-native-application/RudrX/issues/161. Based on the convention that the name of application is the same as that of a workload,  so the pod name will be `APPNAME.*`. For example,
appName: label-example
podName: label-example-6bbd8c6b5b-ssfvx

And this solution can also tail pod logs if users forget to set labels for a pod (https://github.com/crossplane/oam-kubernetes-runtime/issues/184).
I will continue with this PR by leveraging of component labels.

## An example for the command.
```
✗ go run cmd/vela/main.go logs -h
tail pods logs of an application

Usage:
  vela logs <APP-NAME> [flags]

Flags:
  -h, --help   help for logs

Global Flags:
  -e, --env string   specify env name for application


✗ go run cmd/vela/main.go logs label-example
+ label-example-6bbd8c6b5b-ssfvx › wordpress
+ label-example-6bbd8c6b5b-cnwhw › wordpress
+ label-example-6bbd8c6b5b-jw6s4 › wordpress
label-example-6bbd8c6b5b-ssfvx wordpress WordPress not found in /var/www/html - copying now...
label-example-6bbd8c6b5b-ssfvx wordpress Complete! WordPress has been successfully copied to /var/www/html
label-example-6bbd8c6b5b-ssfvx wordpress AH00558: apache2: Could not reliably determine the server's fully qualified domain name, using 10.244.0.7. Set the 'ServerName' directive globally to suppress this message
label-example-6bbd8c6b5b-ssfvx wordpress AH00558: apache2: Could not reliably determine the server's fully qualified domain name, using 10.244.0.7. Set the 'ServerName' directive globally to suppress this message
label-example-6bbd8c6b5b-ssfvx wordpress [Wed Aug 26 11:38:34.444374 2020] [mpm_prefork:notice] [pid 1] AH00163: Apache/2.4.38 (Debian) PHP/7.2.33 configured -- resuming normal operations
label-example-6bbd8c6b5b-ssfvx wordpress [Wed Aug 26 11:38:34.445105 2020] [core:notice] [pid 1] AH00094: Command line: 'apache2 -D FOREGROUND'
label-example-6bbd8c6b5b-cnwhw wordpress WordPress not found in /var/www/html - copying now...
label-example-6bbd8c6b5b-cnwhw wordpress Complete! WordPress has been successfully copied to /var/www/html
label-example-6bbd8c6b5b-cnwhw wordpress AH00558: apache2: Could not reliably determine the server's fully qualified domain name, using 10.244.0.9. Set the 'ServerName' directive globally to suppress this message
label-example-6bbd8c6b5b-cnwhw wordpress AH00558: apache2: Could not reliably determine the server's fully qualified domain name, using 10.244.0.9. Set the 'ServerName' directive globally to suppress this message
label-example-6bbd8c6b5b-cnwhw wordpress [Wed Aug 26 11:38:34.449425 2020] [mpm_prefork:notice] [pid 1] AH00163: Apache/2.4.38 (Debian) PHP/7.2.33 configured -- resuming normal operations
label-example-6bbd8c6b5b-cnwhw wordpress [Wed Aug 26 11:38:34.449850 2020] [core:notice] [pid 1] AH00094: Command line: 'apache2 -D FOREGROUND'
label-example-6bbd8c6b5b-jw6s4 wordpress WordPress not found in /var/www/html - copying now...
label-example-6bbd8c6b5b-jw6s4 wordpress Complete! WordPress has been successfully copied to /var/www/html
label-example-6bbd8c6b5b-jw6s4 wordpress AH00558: apache2: Could not reliably determine the server's fully qualified domain name, using 10.244.0.4. Set the 'ServerName' directive globally to suppress this message
label-example-6bbd8c6b5b-jw6s4 wordpress AH00558: apache2: Could not reliably determine the server's fully qualified domain name, using 10.244.0.4. Set the 'ServerName' directive globally to suppress this message
label-example-6bbd8c6b5b-jw6s4 wordpress [Wed Aug 26 11:38:34.431933 2020] [mpm_prefork:notice] [pid 1] AH00163: Apache/2.4.38 (Debian) PHP/7.2.33 configured -- resuming normal operations
label-example-6bbd8c6b5b-jw6s4 wordpress [Wed Aug 26 11:38:34.432103 2020] [core:notice] [pid 1] AH00094: Command line: 'apache2 -D FOREGROUND'
```

# Questions
To use source code or binary of https://github.com/wercker/stern, which is better?

As the maintainer of stern left the company which supported stern, there has been no commits more than one year. Based on the forked project https://github.com/grosser/stern and make some modifications (see latest commits https://github.com/zzxwill/stern/commits/master) to finally make it integrable in source code level by vela.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -18,9 +19,9 @@ require (
 	github.com/spf13/cobra v1.0.0
 	github.com/spf13/pflag v1.0.5
 	github.com/stretchr/testify v1.6.1
+	github.com/zzxwill/stern v1.13.3-alpha
```
Comment (wonderflow) [inline go.mod:22]: please use https://github.com/oam-dev/stern, and keep the old lib like this : https://github.com/rancher/rio/blob/master/go.mod#L18

File: go.mod [hunk 2]
```diff
@@ -4,6 +4,7 @@ go 1.13
 
 require (
 	cuelang.org/go v0.2.2
+	github.com/Azure/go-autorest v12.2.0+incompatible // Don't remove. https://github.com/kubernetes/client-go/issues/628
```
Comment (wonderflow) [inline go.mod:7]: Can it be fixed by https://github.com/kubernetes/client-go/issues/628#issuecomment-569683255 ?

Comment (zzxwill) [reply go.mod:7]: No. These two lines doesn't show in go.mod.
```
github.com/Azure/go-autorest/autorest v0.9.2 // indirect
github.com/Azure/go-autorest/autorest/to v0.3.0 // indirect
```
They are not direct dependency.

Comment (wonderflow) [reply go.mod:7]: ok

**Comment by zzxwill** [comment]
## vela help grouping message.
grouped `vela logs` to `Applications`
```
➜  /Users/zhouzhengxi/Programming/golang/src/github.com/zzxwill/RudrX git:(log) ✗ go run cmd/vela/main.go -h
✈️  A Micro App Platform for Kubernetes.

Usage:
  vela [flags]
  vela [command]

Available Commands:

  Getting Started:
    env    	Manage application environments
    version	Prints out build version information

  Applications:
    app           	Manage applications with ls, show, delete, run
    logs <appname>	Tail pods logs of an application

  Workloads:
    comp <commands>	Manage Components

    Want more? < install more capabilities by `vela cap` >

  Traits:
    manualscaler <appname> [args]	Attach manualscaler trait to an app
    rollout <appname> [args]     	Attach rollout trait to an app
    route <appname> [args]       	Attach route trait to an app
    scale <appname> [args]       	Attach scale trait to an app

    Want more? < install more capabilities by `vela cap` >

  Release:


  Others:
    cap <command>	Capability Management with config, list, add, remove capabilities

  System:
    completion < bash | zsh >	Output shell completion code for the specified shell (bash or zsh...
    dashboard                	Setup API Server and launch Dashboard
    system                   	system management utilities

Flags:
  -h, --help   help for vela

Use "vela [command] --help" for more information about a command.
```

## vela logs help
```
```

File: pkg/cmd/logs.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+		return err
+	}
+	tails := make(map[string]*stern.Tail)
+	logC := make(chan string, 1024)
+
+	go func() {
+		for {
+			select {
+			case str := <-logC:
+				ioStreams.Infonln(str)
```
Comment (zzxwill) [inline pkg/cmd/logs.go:70]: The naming of function `Infonln()` is a little bit wired. How about directly using `fmt.Fprintf(os.Stdout, str)` instead?

File: pkg/cmd/logs.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+		}
+		return nil
+	}
+	cmd.Annotations = map[string]string{
+		types.TagCommandType: types.TypeApp,
+	}
+	return cmd
+}
+
+// Run refer to the implementation at https://github.com/oam-dev/stern/blob/master/stern/main.go
```
Comment (zzxwill) [inline pkg/cmd/logs.go:51]: This is the reference link for this implementation.

Comment (wonderflow) [reply pkg/cmd/logs.go:51]: refer to the original link : github.com/wercker/stern

File: pkg/cmd/logs.go (hunk 3 above)
Comment (wonderflow) [reply pkg/cmd/logs.go:70]: we'd better align with the whole project, you can try to refactor `cmdutil.IOStreams`

**Review by wonderflow — APPROVED** [review]
Excellent job!

File: pkg/cmd/logs.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+	cmd := &cobra.Command{}
+	cmd.Use = "logs <appname>"
+	cmd.Short = "Tail pods logs of an application"
+	cmd.Long = "Tail pods logs of an application"
+	cmd.PreRunE = func(cmd *cobra.Command, args []string) error {
+		if len(args) < 1 {
+			ioStreams.Errorf("please specify the application name")
+		}
+		appName := args[0]
+		cmd.SetArgs([]string{appName + ".*"})
```
Comment (wonderflow) [inline pkg/cmd/logs.go:25]: This is hard to read

File: pkg/cmd/logs.go (hunk 4 above)
Comment (zzxwill) [reply pkg/cmd/logs.go:51]: Done in the latest commit. Thanks.

**Pull request merged**